File: src/api/src/todo.ts
import { Status, Todo, User } from '../../core'
import { Request } from '../axios/index'
import { ApiResponse } from './type'

// 使用Request类的init方法初始化axios实例
const request = Request.init()

/**
 * 添加新的待办事项
 * @param todo 待添加的待办事项对象
 * @returns 返回包含用户信息的ApiResponse对象
 */
export const addNewTodo = async (todo: Todo): Promise<ApiResponse<User>> => {
  const { data } = await request.post('/todo/create', todo)
  return data
}

/**
 * 删除待办事项
 * @param username 用户名
 * @param id 待办事项的ID
 * @returns 返回包含用户信息的ApiResponse对象
 */
export const deleteTodo = async (username: string, id: string): Promise<ApiResponse<User>> => {
  const { data } = await request.delete('/todo/' + username + '/' + id)
  return data
}

/**
 * 更新待办事项
 * @param username 用户名
 * @param id 待办事项的ID
 * @param todo 更新后的待办事项对象
 * @returns 返回包含用户信息的ApiResponse对象
 */
export const updateTodo = async (username: string, id: string, todo: Todo): Promise<ApiResponse<User>> => {
  const { data } = await request.put('/todo/' + username + '/' + id, todo)
  return data
}

/**
 * 标记待办事项为失败
 * @param username 用户名
 * @param id 待办事项的ID
 * @param todo 包含失败状态的待办事项对象
 * @returns 返回包含用户信息的ApiResponse对象
 */
export const failedTodo = async (username: string, id: string, todo: Todo): Promise<ApiResponse<User>> => {
  const { data } = await request.put('/todo/failed/' + username + '/' + id, todo)
  return data
}

/**
 * 更新待办事项的状态
 * @param id 待办事项的ID
 * @param status 新的状态值
 * @returns 返回一个布尔值，表示更新状态是否成功
 */
export const updateTodoStatus = async (id: String, status: Status): Promise<boolean> => {
  const { data } = await request.get('/todo/' + id + '/' + status)
  return data
}

/**
 * 标记待办事项为完成
 * @param username 用户名
 * @param todoId 待办事项的ID
 * @returns 返回包含用户信息的ApiResponse对象
 */
export const completedTodo = async (username: string, todoId: string): Promise<ApiResponse<User>> => {
  const { data } = await request.get('/todo/complete/' + username + '/' + todoId)
  return data
}

File: back_end/todo/src/lib/api/todo.rs
use rocket::serde::json::Json;

use crate::lib::{
    entry::{
        po::{Status, Todo, TodoBox},
        dto::{self},
    },
    error::Error,
    mapping::{
        create_todo_by_username, delete_todo_by_id, select_todo_record, select_user_by_username,
        update_todo_by_id, update_todo_status_by_id, update_user_by_username,
    },
    response::ResultJsonData,
};

#[post("/create", format = "application/json", data = "<todo>")]
pub async fn create_todo(todo: Json<Todo>) -> ResultJsonData<dto::User> {
    let todo = todo.0;
    let username = todo.clone().owner;
    let query = create_todo_by_username(&username, todo).await;
    if let Some((id, todo)) = query {
        let mut user = select_user_by_username(&username).await.unwrap();
        let priority = todo.priority();
        let _ = user.add_todo(id, priority);
        //update user
        let update_query = update_user_by_username(user).await;
        if let Some(user) = update_query {
            let user = dto::User::from(user).await;
            return ResultJsonData::success(user);
        } else {
            let e = Error::UpdateUser;
            let (e_code, e_msg) = e.get();
            return ResultJsonData::define_failure(e_code, &e_msg);
        }
    }
    let e = Error::CreateTodo;
    let (e_code, e_msg) = e.get();
    return ResultJsonData::define_failure(e_code, &e_msg);
}

#[delete("/<username>/<id>")]
pub async fn delete_todo(username: &str, id: &str) -> ResultJsonData<dto::User> {
    let query = delete_todo_by_id(id).await;
    if query {
        let user_query = select_user_by_username(username).await;
        if let Some(mut user) = user_query {
            let _ = user.delete_todo(id);
            let update_query = update_user_by_username(user).await;
            if let Some(user) = update_query {
                let user = dto::User::from(user).await;
                return ResultJsonData::success(user);
            }
        }
    }
    let e = Error::DeleteTodo;
    let (e_code, e_msg) = e.get();
    return ResultJsonData::define_failure(e_code, &e_msg);
}

#[put("/<username>/<id>", format = "application/json", data = "<todo>")]
pub async fn update_todo(username: &str, id: &str, todo: Json<Todo>) -> ResultJsonData<dto::User> {
    let query = update_todo_by_id(&id, todo.0).await;

    if query {
        let user_query = select_user_by_username(username).await;
        if let Some(mut user) = user_query {
            let user = dto::User::from(user).await;
            return ResultJsonData::success(user);
        }
    }
    let e = Error::UpdateTodo;
    let (e_code, e_msg) = e.get();
    return ResultJsonData::define_failure(e_code, &e_msg);
}

#[put(
    "/failed/<username>/<id>",
    format = "application/json",
    data = "<todo>"
)]
pub async fn failed_todo(username: &str, id: &str, todo: Json<Todo>) -> ResultJsonData<dto::User> {
    let query = update_todo_by_id(&id, todo.0).await;
    if query {
        let user_query = select_user_by_username(username).await;
        if let Some(mut user) = user_query {
            let _ = user.failed_todo(id);
            let query = update_user_by_username(user).await;
            if let Some(user) = query {
                let user = dto::User::from(user).await;
                return ResultJsonData::success(user);
            }
        }
    }
    let e = Error::UpdateTodo;
    let (e_code, e_msg) = e.get();
    return ResultJsonData::define_failure(e_code, &e_msg);
}

#[get("/<id>/<status>")]
pub async fn update_todo_status(id: &str, status: &str) -> ResultJsonData<bool> {
    let query = update_todo_status_by_id(id, status).await;
    if query {
        return ResultJsonData::success(query);
    }
    let e = Error::UpdateTodo;
    let (e_code, e_msg) = e.get();
    return ResultJsonData::define_failure(e_code, &e_msg);
}

#[get("/complete/<username>/<id>")]
pub async fn complete_todo(username: &str, id: &str) -> ResultJsonData<dto::User> {
    let status = Status::Completed.to_string();
    let (code, _update_status) = update_todo_status(id, &status).await.get();
    if code == 200 {
        let user_query = select_user_by_username(username).await;
        if let Some(mut user) = user_query {
            let _ = user.complete_todo(id);
            let query = update_user_by_username(user).await;
            if let Some(user) = query {
                let user = dto::User::from(user).await;
                return ResultJsonData::success(user);
            }
        }
    }
    let e = Error::CompleteTodo;
    let (e_code, e_msg) = e.get();
    return ResultJsonData::define_failure(e_code, &e_msg);
}
